File: packages/guards/src/rag-safety.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface RagSafetyOptions {
  action: 'block' | 'warn';
  /** Maximum allowed context-to-answer ratio */
  maxContextRatio?: number;
  /** Check for context manipulation */
  checkManipulation?: boolean;
}

const CONTEXT_MANIPULATION_PATTERNS: RegExp[] = [
  /\b(ignore|disregard|override)\s+(the\s+)?(context|document|source|retrieved)/i,
  /\b(context|document)\s+(says?\s+to\s+)?(ignore|override|forget)/i,
  /\bpretend\s+(the\s+)?(context|document|source)\s+(says?|contains?|mentions?)/i,
  /\bmake\s+up\s+(an?\s+)?answer/i,
  /\b(fabricate|invent|hallucinate)\s+(a\s+)?response/i,
  /\bdon'?t\s+use\s+(the\s+)?(context|documents?|sources?)/i,
];

const CONTEXT_POISONING_PATTERNS: RegExp[] = [
  /\[hidden\s*:\s*/i,
  /<!--\s*(instruction|command|override)/i,
  /\bzero[\s-]?width\s+(space|char)/i,
  /\u200B|\u200C|\u200D|\uFEFF/,
  /\bdata:text\/html/i,
  /\bjavascript:/i,
];

const ATTRIBUTION_PATTERNS: RegExp[] = [
  /\baccording\s+to\s+(no|fake|made[\s-]?up)\s+source/i,
  /\bsource:\s*(none|n\/a|unavailable|unknown)\b/i,
];

export function ragSafety(options: RagSafetyOptions): Guard {
  const checkManipulation = options.checkManipulation ?? true;

  return {
    name: 'rag-safety',
    version: '0.1.0',
    description: 'RAG pipeline safety: context manipulation, poisoning, zero-width injection, attribution check',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const violations: string[] = [];

      if (checkManipulation) {
        for (const p of CONTEXT_MANIPULATION_PATTERNS) {
          if (p.test(text)) { violations.push('context-manipulation'); break; }
        }
      }

      for (const p of CONTEXT_POISONING_PATTERNS) {
        if (p.test(text)) { violations.push('context-poisoning'); break; }
      }

      for (const p of ATTRIBUTION_PATTERNS) {
        if (p.test(text)) { violations.push('attribution-issue'); break; }
      }

      const triggered = violations.length > 0;

      return {
        guardName: 'rag-safety',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min(violations.length / 2, 1.0) : 0,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { violations } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/rag_safety.py
"""RAG pipeline safety guard."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_MANIPULATION_PATTERNS: list[re.Pattern[str]] = [
    re.compile(
        r"(ignore|disregard)\s+the\s+context",
        re.IGNORECASE,
    ),
    re.compile(
        r"make\s+up\s+an\s+answer",
        re.IGNORECASE,
    ),
    re.compile(
        r"fabricate\s+response",
        re.IGNORECASE,
    ),
]

_POISONING_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"\[hidden:", re.IGNORECASE),
    re.compile(r"<!--.*?-->"),
    re.compile(r"[\u200b\u200c\u200d\ufeff]"),
    re.compile(r"data:text/html", re.IGNORECASE),
    re.compile(r"javascript:", re.IGNORECASE),
]

_ATTRIBUTION_PATTERNS: list[re.Pattern[str]] = [
    re.compile(
        r"according\s+to\s+(no|fake)\s+source",
        re.IGNORECASE,
    ),
    re.compile(
        r"source:\s*(none|unavailable)",
        re.IGNORECASE,
    ),
]


class _RagSafety:
    def __init__(
        self,
        *,
        action: str = "block",
        check_manipulation: bool = True,
        extra_patterns: Optional[List[re.Pattern[str]]] = None,
    ) -> None:
        self.name = "rag-safety"
        self.action = action
        self._patterns: list[re.Pattern[str]] = []
        if check_manipulation:
            self._patterns.extend(_MANIPULATION_PATTERNS)
        self._patterns.extend(_POISONING_PATTERNS)
        self._patterns.extend(_ATTRIBUTION_PATTERNS)
        if extra_patterns:
            self._patterns.extend(extra_patterns)

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in self._patterns:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = min(len(matched) / 3, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="rag-safety",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "RAG safety violation detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "matched_patterns": len(matched),
                    "reason": (
                        "Text contains patterns indicating"
                        " RAG pipeline manipulation,"
                        " context poisoning, or"
                        " attribution issues"
                    ),
                }
                if triggered
                else None
            ),
        )


def rag_safety(
    *,
    action: str = "block",
    check_manipulation: bool = True,
    extra_patterns: Optional[List[re.Pattern[str]]] = None,
) -> _RagSafety:
    return _RagSafety(
        action=action,
        check_manipulation=check_manipulation,
        extra_patterns=extra_patterns,
    )
